src/network/core/http_winhttp.cpp

```diff
@@ -0,0 +1,301 @@
+			}
+
+			break;
+
+		case WINHTTP_CALLBACK_STATUS_SECURE_FAILURE:
+		case WINHTTP_CALLBACK_STATUS_REQUEST_ERROR:
+			Debug(net, 0, "HTTP request failed: {}", GetLastErrorAsString());
+			this->callback->OnFailure();
+			this->finished = true;
+			break;
+
+		default:
+			Debug(net, 0, "HTTP request failed: unexepected callback code 0x{:x}", code);
+			this->callback->OnFailure();
+			this->finished = true;
+			return;
+	}
+}
+
+static void CALLBACK StaticWinHttpCallback(HINTERNET handle, DWORD_PTR context, DWORD code, void *info, DWORD length)
+{
+	NetworkHTTPRequest *request = (NetworkHTTPRequest *)context;
+	request->WinHttpCallback(code, info, length);
+}
+
+/**
+ * Start the HTTP request handling.
+ *
+ * This is done in an async manner, so we can do other things while waiting for
+ * the HTTP request to finish. The actual receiving of the data is done in
+ * Receive().
+ */
+void NetworkHTTPRequest::Connect()
+{
+	Debug(net, 1, "HTTP request to {}", std::string(uri.begin(), uri.end()));
+
+	URL_COMPONENTS url_components = {};
+	wchar_t scheme[32];
+	wchar_t hostname[128];
+	wchar_t url_path[4096];
+
+	/* Convert the URL to its components. */
+	url_components.dwStructSize = sizeof(url_components);
+	url_components.lpszScheme = scheme;
+	url_components.dwSchemeLength = lengthof(scheme);
+	url_components.lpszHostName = hostname;
+	url_components.dwHostNameLength = lengthof(hostname);
+	url_components.lpszUrlPath = url_path;
+	url_components.dwUrlPathLength = lengthof(url_path);
+	WinHttpCrackUrl(this->uri.c_str(), 0, 0, &url_components);
+
+	/* Create the HTTP connection. */
+	this->connection = WinHttpConnect(_winhttp_session, url_components.lpszHostName, url_components.nPort, 0);
+	assert(this->connection != nullptr);
+	this->request = WinHttpOpenRequest(connection, data == nullptr ? L"GET" : L"POST", url_components.lpszUrlPath, nullptr, WINHTTP_NO_REFERER, WINHTTP_DEFAULT_ACCEPT_TYPES, url_components.nScheme == INTERNET_SCHEME_HTTPS ? WINHTTP_FLAG_SECURE : 0);
+	assert(this->request != nullptr);
+
+	/* Send the request (possibly with a payload). */
+	if (data == nullptr) {
+		WinHttpSendRequest(this->request, WINHTTP_NO_ADDITIONAL_HEADERS, 0, WINHTTP_NO_REQUEST_DATA, 0, 0, reinterpret_cast<DWORD_PTR>(this));
+	} else {
+		WinHttpSendRequest(this->request, L"Content-Type: application/x-www-form-urlencoded\r\n", -1, const_cast<char *>(data), static_cast<DWORD>(strlen(data)), static_cast<DWORD>(strlen(data)), reinterpret_cast<DWORD_PTR>(this));
+	}
+}
+
+/**
+ * Poll and process the HTTP request/response.
+ *
+ * @return True iff the request is done; no call to Receive() should be done after it returns true.
+ */
+bool NetworkHTTPRequest::Receive()
+{
+	return this->finished;
+}
+
+/**
+ * Destructor of the HTTP request.
+ *
+ * Makes sure all handlers are closed, and all memory is free'd.
+ */
+NetworkHTTPRequest::~NetworkHTTPRequest()
+{
+	if (this->request) {
+		WinHttpCloseHandle(this->request);
+		WinHttpCloseHandle(this->connection);
```

Comment:
Technically leaks resources (in release builds without asserts) in case the `connection` is opened by the `request` isn't.

Reply:
I was lazy, and you are right :P Let me fix it properly :)